File: julia-parallel/matrix-free-tensor/stokes_flow/afunstreambc.jl
###AFUNSTREAMBC
#input
    # u           vector to multiply
    # xy          Q2 nodal coordinate vector
    # xyp         Q1 nodal coordinate vector
    # mv          Q2 element mapping matrix
    # bound       indices of boundary points
    # ae          local Q2 diffusion derivative matrix
#output
    # w           A * u
function afunstreambc(u, kparams)

  xy = kparams["xy"]; xyp = kparams["xyp"]; mv = kparams["mv"]; bound = kparams["bound"];
  ae = kparams["ae"]

  # get variables
  nvtx = length(xy[:, 1]); nu = 2nvtx; np = 3length(xyp[:, 1])
  nel = length(mv[:, 1])
  aes = squeeze(ae[1, :, :], 1)
  w = zeros(nu+np)
  # w = SharedArray(Float64, (nvtx, 1), pids = workers())

  # zero dirichlet boundary conditions
  uu = copy(u)
  u[bound] = zeros(length(bound))

  n,m = size(mv)
  U = zeros(m, n)

  # @parallel for e = 1:nel
  for e = 1:nel
    ind = vec(mv[e, :]')
    U[:, e] = u[ind]
  end
  W = aes * U
  for e = 1:nel
    ind = vec(mv[e, :]')
    w[ind] += W[:, e]
  end

  w[bound] = uu[bound]
  vec(w)
end

File: matlab/matrix-free-stacked/stokes_flow/afunstreambc.m
function w = afunstreambc(u, xy, xyp, mv, bound, ae)

x = xy(:, 1);
nvtx = length(x);
nel = length(mv(:, 1));

aes = squeeze(ae(1, :, :));

w = zeros(nvtx, 1);
for e = 1:nel
    ind = mv(e, :);
    indbd = ismember(ind, bound);
    indint = ~ismember(ind, bound);
    indb = ind(indbd);

    ux = u(ind);
    ux(indbd) = zeros(1, length(find(indbd)));
    wex = aes * ux;

    % impose boundary conditions
    wex(indbd) = u(indb);
    
    % add interior points
    w(ind(indint)) = w(ind(indint)) + wex(indint);

    % add boundary points only when entry == 0
    wb = w(indb);
    ind0 = find(wb == 0);
    wexb = wex(indbd);
    w(indb(ind0)) = w(indb(ind0)) + wexb(ind0);
end

end